src/modules/submodules/consensus.py

```diff
@@ -82,82 +110,268 @@ def _get_member_info(self, blockstamp: BlockStamp) -> MemberInfo:
+        }
+        """
+        consensus_contract = self._get_consensus_contract(blockstamp)
+        return consensus_contract.functions.getFrameConfig().call(
+            block_identifier=blockstamp.block_hash,
+        )
+
+    # ----- Calculation reference slot for report -----
+
     def get_blockstamp_for_report(self, blockstamp: BlockStamp) -> Optional[BlockStamp]:
+        """Get blockstamp that should be used to build and send report for current frame."""
         member_info = self._get_member_info(blockstamp)
 
-        if blockstamp.slot_number < member_info.member_ref_slot:
+        latest_blockstamp = self._get_latest_blockstamp()
+
+        if not self.is_contract_reportable(latest_blockstamp):
+            logger.info({'msg': 'Contract is not reportable.'})
+            return
+
+        if latest_blockstamp.slot_number < member_info.current_frame_ref_slot:
             logger.info({'msg': 'Reference slot is not yet finalized.'})
             return
 
-        # Maybe check head slot number?
-        if blockstamp.slot_number > member_info.deadline_slot:
+        if not member_info.is_fast_line:
+            if latest_blockstamp.slot_number + member_info.fast_line_length_slot < member_info.current_frame_ref_slot:
+                logger.info({'msg': f'Member is not in fast lane, so report will be postponed for [{member_info.fast_line_length_slot}] slots.'})
+                return
+
+        if latest_blockstamp.slot_number > member_info.deadline_slot:
             logger.info({'msg': 'Deadline missed.'})
             return
 
-        return self._get_first_non_missed_slot(blockstamp, member_info.current_ref_slot)
+        return self._get_first_non_missed_slot(blockstamp, member_info.current_frame_ref_slot)
 
     def _get_first_non_missed_slot(self, blockstamp: BlockStamp, slot: SlotNumber) -> BlockStamp:
-        _, epoch_per_frame = self._get_frame_config(blockstamp)
+        _, epoch_per_frame = self._get_current_frame(blockstamp)
 
         for i in range(slot, slot - epoch_per_frame * 32, -1):
             try:
                 root = self.w3.cc.get_block_root(i).root
-            except KeyError:
-                logger.warning({'msg': f'Missed slot: {i}. Check next slot.'})
+            except NotOkResponse as error:
+                if error.status != HTTPStatus.NOT_FOUND:
+                    raise error from error
+
+                logger.warning({'msg': f'Missed slot: {i}. Check next slot.', 'error': error})
                 continue
-            except NotOkResponse as e:
-                if 'Response [404]' in e.args[0]:
-                    logger.warning({'msg': f'Missed slot: {i}. Check next slot.'})
-                    continue
+            else:
+                slot_details = self.w3.cc.get_block_details(root)
+
+                execution_data = slot_details.message.body['execution_payload']
+
+                return BlockStamp(
+                    block_root=root,
+                    slot_number=slot,
```

Comment:
Shouldn't it be a slot number from `slot_details`?
For now, if this slot is missed, we will get them here